models/register.php

```diff
@@ -2,13 +2,127 @@
     ];
+
+    public function Action(){
+        // この中で登録処理を呼び出す
+        if (isset($_POST["insert"])){
+            // 入力チェック
+            $input_check = new Account();
+            $username_message = $input_check->Get_account_name_errorMessage();
+            $email_message = $input_check->Get_email_errorMessage();
+            $password_message = $input_check->Get_password_errorMessage();
+            if($username_message || $email_message || $password_message){
+                $this->page_data['username_errorMessage'] = $username_message;
+                $this->page_data['email_errorMessage'] = $email_message;
+                $this->page_data['password_errorMessage'] = $password_message;
+            }else{
+                $input_check->Insert_Accounts();
+                $errormessage = $input_check->Get_errorMessage();
+                if($errormessage){
+                    echo "<script>alert('$errormessage');</script>";
+                }
+            }
+        }
+    }
+
+}
+
+class Account {
+    private $account_name;              // ユーザ名
+    private $email;                     // メールアドレス
+    private $password;                  // パスワード
+    private $errorflg;                  // エラーフラグ
+    private $errorMessage;              // 汎用エラー
+    private $account_name_errorMessage; // ユーザ名エラーメッセージ
+    private $email_errorMessage;        // メールアドレスエラーメッセージ
+    private $password_errorMessage;     // パスワードエラーメッセージ
+
+    function __construct(){
+        // 画面から送信された内容を取得
+        $this->account_name = $_POST['account_name'];
+        if(!isset($this->account_name)){
+            $this->account_name = '';
+        }
+        $this->email = $_POST['email'];
+        if(!isset($this->email)){
+            $this->email = '';
+        }
+        $this->password = $_POST['password'];
+        if(!isset($this->password)){
+            $this->password = '';
+        }
+        $this->Input_check();
+    }
+
+    function Insert_Accounts(){
+        try{
+            //データベース接続
+            $DB = new DB_Connect();
+            $pdo = $DB->getPDO();
+            // SQL準備
+            $stmt = $pdo->prepare('INSERT INTO Accounts (account_name, email, password_hash)VALUES(:account_name, :email, :password_hash)');
+            // プレースホルダの値をセット
+            $password_hash = password_hash($this->password, PASSWORD_BCRYPT);
+            $stmt->bindParam(':account_name',$this->account_name, PDO::PARAM_STR);
+            $stmt->bindParam(':email', $this->email, PDO::PARAM_STR);
+            $stmt->bindParam(':password_hash', $password_hash, PDO::PARAM_STR);
+            // ＳＱＬ実行部分
+            $check = $stmt->execute();
+            if($check){
+                echo "<script>alert('アカウントの登録に成功しました！');</script>";
+            }else{
+                echo "<script>alert('アカウントの登録に失敗しました。');</script>";
+            }
+        }catch(PDOException $e){
+            $this->errorMessage .= 'データベース更新エラー\n'.addslashes($e->getMessage()).'。';

```

Comment:
特に自作していないモジュールの例外をcatchする時は要注意です！

当然ですが、$eの中にはPDOException内で定義されたエラーメッセージが入っています。
しかし、そのエラーメッセージは、このサービスに対しては何の配慮もない内容です。
今回の場合、UNIQUEキー重複が発生すると、DBに実際に定義されているカラム名が表示されます。
エラーによってはDBの内容が類推可能となる内容が出力されてしまう可能性もあり、攻撃を受ける起点になる事もあります。

今回の場合は、画面には「アカウントの登録に失敗しました」と出力しておき、PDOExceptionからのエラーメッセージはphpのエラーログに出力するのが良いでしょう。

```suggestion
            error_log("[". date('Y-m-d H:i:s') . "]アカウント登録エラー：".addslashes($e->getMessage())."\n", 3, "/var/log/php/php_error.log");
            $this->errorMessage .= 'アカウントの登録に失敗しました。\nサイトの管理者に連絡してください。';
```

Reply:
了解しました、気を付けます。
今回はログ保存と表示を修正しておきます。

Reply:
ＯＫです。直ってるの確認しました。